### ISSUE TITLE
Virtual keyboard navigation on NT

### ISSUE THREAD

**Issue by Commodore64user** [koreader/koreader#12100]
* KOReader version: Sourdough
* Device: Non Touch in general

#### Issue
When using a virtual keyboard on non-touch devices, navigation can sometimes feel like a guessing game. For instance, consider the following scenarios:

If you focus on the `Enter` key and press the `Up` key, you’d expect the focus to move to the `Del` key. However, it ends up going to the `N` key instead. (picture for reference)
similarly, if you position yourself on the `7` key and press `Up`, you’d anticipate looping to the `space bar`. Unfortunately, it takes you to the `Enter` key, which can be quite confusing.

<img src="https://github.com/koreader/koreader/assets/97603719/7cae7277-e590-43b1-8748-0a4d2ef5a75b" width=50% height=50%>

I took the liberty of making this _**non-exhaustive**_ table of the current behaviour going downwards from row 4 to row 5, and what I think would be better.

| From    | Goes to          | where it should go to    |
| ------- | ------------- | ---------------------- |
| `Shift` | `Symbol`       | `Symbol`                      |
| `Z`       | `Globe`         | `Globe `                       |
| `X`       | `.`                  | `.`                                 |
| `C`       | `Space`        | `Space `                      |
| `V`       | `left arrow`   | `Space `                      |
| `B`       | `right arrow` | `Space `                      |
| `N`       | `Enter`          | `left arrow`                  |
| `M`       | `Enter`          | `right arrow`               |
| `Del`     | `Enter`         | `Enter `                        |

Note: I do not want to touch this cluster bomb as it appears all other keyboard layouts are dependant on the English one.

**Comment by Frenzie** [comment]
They're both number 7. You'd need more advanced positioning info.

**Comment by Commodore64user** [comment]
> They're both number 7. You'd need more advanced positioning info.

yes, but that is not visually intuitive is I guess my point.

**Comment by NiLuJe** [comment]
Yep, FocusManager's simple grid array sort of assumes the rows/columns are mostly identically sized, so that when going up/down, the current column roughly matches, and vice-versa.

The VK... very much breaks that assumption (and it's probably very layout-dependent, too, although we *may* have made some effort to keep the layout similar, I don't recall).

That would basically require FocusManager to somehow deal with absolute coordinates, and be able to say if I go up, is there a widget in my layout at these coordinates, and at where is it in my layout grid?

That's... a nightmarish undertaking ;).

**Comment by Commodore64user** [comment]
So as with a lot of things in life, high effort, low reward. got it.
